applications/nrf_desktop/src/util/hid_reportq.c

```diff
@@ -0,0 +1,295 @@
+/*
+ * Copyright (c) 2024 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-Nordic-5-Clause
+ */
+
+#include <stdint.h>
+#include <zephyr/sys/slist.h>
+#include <zephyr/kernel.h>
+
+#include "hid_reportq.h"
+#include "hid_report_desc.h"
+#include "hid_event.h"
+
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(hid_reportq, CONFIG_DESKTOP_HID_REPORTQ_LOG_LEVEL);
+
+#define MAX_ENQUEUED_REPORTS CONFIG_DESKTOP_HID_REPORTQ_MAX_ENQUEUED_REPORTS
+
+struct enqueued_report {
+	sys_snode_t node;
+	struct hid_report_event *event;
+};
+
+struct counted_list {
+	sys_slist_t list;
+	size_t node_count;
+};
+
+struct hid_reportq {
+	struct counted_list reports[ARRAY_SIZE(input_reports)];
+	uint16_t enabled_report_idx_bm;
+	uint8_t last_sent_report_idx;
+	uint8_t report_max;
+	uint8_t report_cnt;
+	const void *sub_id;
```

Comment:
There can be only one subscriber per queue (subscriber is HID report sink; in current use-case subscriber is a USB HID class instance). Current approach allows to properly track subscriptions, queue HID reports and limit number of HID reports provided to a given USB HID class instance at a time. A queue object can be shared among multiple data sources (so that multiple sources can add reports to the same queue).

I am not sure how could this work for multiple subscribers. I would suggest to discuss offline if needed.